# AMSC698K Homework 10
##### Elijah Kin & Noorain Noorani

In [1]:
%%capture

# Run VQE notebook to get Pauli decomposition function
%run VQE_AHO.ipynb

In [2]:
import numpy as np

### Problem A
1. The Hadamard Test circuit takes a qubit $q_0$, prepared as $|+\rangle$, and state $|\psi\rangle \in \mathbb{C}^n$ and applies to it the controlled unitary CU. Describe how this test works and what the measurement of $q_0$ tells us about $U |\psi\rangle$. 

2. In the VQLS algorithm the Hadamard Test is applied in a modified form since two controlled unitaries are provided for the test. Describe how this test results in an "overlap probability" (or cosine similarity, a central measure in NLP).

In [3]:
# TODO

### Problem B
1. Suppose a LSE, decomposed into quantum gates, is given by:
$$ A = \mathbb{I} + 0.2 X_0 Z_1 + 0.2 X_0; \quad |b\rangle = H_0 H_1 H_2 |0\rangle $$
Apply the VQLS algorithm and compare with classical solutions using $\texttt{scipy.linalg.solve()}$ and $\texttt{scipy.linalg.inv()}$.

In [4]:
# TODO

2. (extra credit) Solve this problem using the HHL algorithm.

In [5]:
# TODO

### Problem C
Given the inhomogeneous linear system:
$$ \begin{align*} x_1 - x_2 + x_3 - x_4 &= 2 \\ x_1 - x_2 + x_3 + x_4 &= 0 \\ 4x_1 - 4x_2 + 4x_3 &= 4 \\ 2x_1 - 2x_2 + 2x_3 - x_4 &= 3 \end{align*} $$
Write this LSE in matrix form and decompose the coefficient matrix $A$ and vector $b$ to apply the VQLS algorithm and compare its result with the "classical" solutions.

In [6]:
# Set up the matrix A and vector b to find their Pauli decompositions
A = np.array([[1, -1, 1, -1],
              [1, -1, 1, 1],
              [4, -4, 4, 0],
              [2, -2, 2, -1]])
print(decompose_H_to_string(A))

b = np.array([[2],
              [0],
              [4],
              [3]])
print(decompose_H_to_string(b))

['0.75 * I^I', '0.5 * I^X', '-1j * I^Y', '1.75 * I^Z', '1 * X^I', '-0.5 * X^X', '-2j * X^Y', '1.5 * X^Z', '0.5j * Y^X', '-1 * Y^Y', '-1.5j * Y^Z', '-0.75 * Z^I', '-0.5 * Z^X', '-0.75 * Z^Z']
['0.5 * I^I', '0.5 * I^Z', '1 * X^I', '0.75 * X^X', '-0.75j * X^Y', '1 * X^Z', '-1j * Y^I', '-0.75j * Y^X', '-0.75 * Y^Y', '-1j * Y^Z', '0.5 * Z^I', '0.5 * Z^Z']


Using $\texttt{decompose\_Pauli}$ from $\texttt{VQE\_AHO.ipynb}$, we find the decompositions
\begin{align*}
  A &= 0.75 \mathbb{I} + 0.5 X_0 - i Y_0 + 1.75 Z_0 + X_1 - 0.5 X_1X_0 - 2i X_1Y_0 + 1.5 X_1Z_0 \\
  & \quad + 0.5i Y_1X_0 - Y_1Y_0 - 1.5i Y_1Z_0 - 0.75 Z_1 - 0.5 Z_1X_0 - 0.5 Z_1Z_0 \\
  b &= 0.5 \mathbb{I} + 0.5 Z_0 + X_1 + 0.75 X_1X_0 - 0.75i X_1Y_0 + X_1Z_0 \\
  & \quad - iY_1 - 0.75iY_1X_0 - 0.75Y_1Y_0 - iY_1Z_0 + 0.5Z_1 + 0.5Z_1Z_0
\end{align*}

In [7]:
# TODO